File: search/array-based-search/self-organizing-search/python/test_self_organizing_search.py
import unittest
from self_organizing_search import SelfOrganizingList


class TestSelfOrganizingList(unittest.TestCase):

    def test_move_to_front_single_access(self):
        sol = SelfOrganizingList(['A', 'B', 'C'], strategy='move_to_front')
        sol.search('B')
        self.assertEqual(sol.items, ['B', 'A', 'C'])

    def test_move_to_front_multiple_access(self):
        sol = SelfOrganizingList(['A', 'B', 'C', 'D'], strategy='move_to_front')
        sol.search('C')  # → [C, A, B, D]
        sol.search('B')  # → [B, C, A, D]
        self.assertEqual(sol.items, ['B', 'C', 'A', 'D'])

    def test_transpose_once(self):
        sol = SelfOrganizingList(['A', 'B', 'C'], strategy='transpose')
        sol.search('C')  # swaps with B → [A, C, B]
        self.assertEqual(sol.items, ['A', 'C', 'B'])

    def test_transpose_multiple(self):
        sol = SelfOrganizingList(['A', 'B', 'C', 'D'], strategy='transpose')
        sol.search('D')  # → [A, B, C, D] (no change if last)
        sol.search('C')  # → [A, B, D, C]
        sol.search('D')  # → [A, B, D, C] (now D is before C)
        self.assertEqual(sol.items, ['A', 'B', 'D', 'C'])

    def test_count_strategy_reorders_correctly(self):
        sol = SelfOrganizingList(['X', 'Y', 'Z'], strategy='count')
        sol.search('Z')
        sol.search('Y')
        sol.search('Z')
        # Z:2, Y:1, X:0 → should be [Z, Y, X]
        self.assertEqual(sol.items, ['Z', 'Y', 'X'])

    def test_count_increments_properly(self):
        sol = SelfOrganizingList(['P', 'Q'], strategy='count')
        sol.search('Q')
        sol.search('Q')
        self.assertEqual(sol.access_count['Q'], 2)
        self.assertEqual(sol.access_count['P'], 0)

    def test_empty_list_search_returns_minus_one(self):
        sol = SelfOrganizingList([], strategy='move_to_front')
        self.assertEqual(sol.search('X'), -1)

    def test_not_found_returns_minus_one(self):
        sol = SelfOrganizingList(['A', 'B'], strategy='move_to_front')
        self.assertEqual(sol.search('C'), -1)

    def test_unaffected_strategies_on_not_found(self):
        sol = SelfOrganizingList(['A', 'B'], strategy='move_to_front')
        sol.search('C')
        self.assertEqual(sol.items, ['A', 'B'])  # Should not change

    def test_case_sensitivity(self):
        sol = SelfOrganizingList(['a', 'A'], strategy='move_to_front')
        self.assertEqual(sol.search('A'), 1)
        self.assertEqual(sol.items, ['A', 'a'])


if __name__ == "__main__":
    unittest.main()

File: search/array-based-search/self-organizing-search/javascript/test_self_organizing_search.js
const { strictEqual, deepEqual } = require('assert');

// Import the class
const { SelfOrganizingList } = require('./self_organizing_search');

console.log("🧪 Running Self-Organizing Search Tests...\n");

// Test: Move-to-Front - single access
(() => {
    const sol = new SelfOrganizingList(['A', 'B', 'C'], 'move_to_front');
    sol.search('B');
    deepEqual(sol.toArray(), ['B', 'A', 'C'], "Should move B to front");
})();

// Test: Move-to-Front - multiple accesses
(() => {
    const sol = new SelfOrganizingList(['A', 'B', 'C', 'D'], 'move_to_front');
    sol.search('C'); // → [C, A, B, D]
    sol.search('B'); // → [B, C, A, D]
    deepEqual(sol.toArray(), ['B', 'C', 'A', 'D'], "Should maintain move-to-front logic");
})();

// Test: Transpose - one swap
(() => {
    const sol = new SelfOrganizingList(['A', 'B', 'C'], 'transpose');
    sol.search('C'); // → [A, C, B]
    deepEqual(sol.toArray(), ['A', 'C', 'B'], "Should swap C with B");
})();

// Test: Transpose - repeated access
(() => {
    const sol = new SelfOrganizingList(['W', 'X', 'Y', 'Z'], 'transpose');
    sol.search('Z'); // no change
    sol.search('Y'); // → [W, X, Z, Y]
    sol.search('Z'); // → [W, X, Z, Y] (Z now before Y)
    deepEqual(sol.toArray(), ['W', 'X', 'Z', 'Y'], "Should improve position over time");
})();

// Test: Count - reorder by frequency
(() => {
    const sol = new SelfOrganizingList(['P', 'Q', 'R'], 'count');
    sol.search('R');
    sol.search('Q');
    sol.search('R');
    // R:2, Q:1, P:0 → [R, Q, P]
    deepEqual(sol.toArray(), ['R', 'Q', 'P'], "Should reorder by count");
    deepEqual(sol.getCounts(), { P: 0, Q: 1, R: 2 }, "Counts should be accurate");
})();

// Test: Count - stable with equal frequencies
(() => {
    const sol = new SelfOrganizingList(['X', 'Y'], 'count');
    sol.search('Y');
    sol.search('X');
    // Both have count 1 → order preserved
    deepEqual(sol.toArray(), ['X', 'Y'], "Equal frequencies should preserve order");
})();

// Test: Empty list returns -1
strictEqual(new SelfOrganizingList([], 'move_to_front').search('X'), -1, "Empty list → -1");

// Test: Not found returns -1 and doesn't modify
(() => {
    const sol = new SelfOrganizingList(['M', 'N'], 'move_to_front');
    strictEqual(sol.search('O'), -1, "Should return -1 for missing element");
    deepEqual(sol.toArray(), ['M', 'N'], "List should remain unchanged");
})();

// Test: Case sensitivity
(() => {
    const sol = new SelfOrganizingList(['a', 'A'], 'move_to_front');
    strictEqual(sol.search('A'), 1, "Should distinguish case");
    deepEqual(sol.toArray(), ['A', 'a'], "A moved to front");
})();

console.log("\n✅ All tests passed!");